### PR TITLE
[BI-998] - Add tags to traits

### PR DESCRIPTION
**Pull request by ctucker3**
Some of the tests fail right now because tags were added and the brapi-server needs to be patched. This can still be review before then, I won't merge until the tests pass. 

Uses the brapi `variable.additionalInfo` field to store the tags. 

Additions:
  - POST traits with tags
  - PUT traits with tags
  - GET all trait tags within a program
  - Add tags to trait upload

Notes:
  - On the front end, existing tags assigned to the trait are based on lower case matching, and removing all spaces before matching. 
  
See other PRs for this feature:
brapi: https://github.com/Breeding-Insight/brapi/pull/40
sgn: https://github.com/Breeding-Insight/sgn/pull/30
bi-web: https://github.com/Breeding-Insight/bi-web/pull/99

LEFT TO DO: 
- Finish up tests in brapi library once Pete fixes the test server

### CODE REVIEW

File: src/main/java/org/breedinginsight/daos/TraitDAO.java [hunk 1]
```diff
@@ -55,13 +56,17 @@
     @Property(name = "brapi.server.reference-source")
     private String referenceSource;
     private ObservationDAO observationDao;
+    private Gson gson;
+
+    private String TAGS_KEY = "tags";
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/daos/TraitDAO.java:61]: could be final static

File: src/main/java/org/breedinginsight/services/parsers/trait/TraitFileColumns.java [hunk 2]
```diff
@@ -36,7 +36,8 @@
     SCALE_DECIMAL_PLACES("Scale decimal places"),
     SCALE_LOWER_LIMIT("Scale lower limit"),
     SCALE_UPPER_LIMIT("Scale upper limit"),
-    SCALE_CATEGORIES("Scale categories");
+    SCALE_CATEGORIES("Scale categories"),
+    TAGS("Tags");
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/services/parsers/trait/TraitFileColumns.java:40]: How is tags different from trait lists?

File: src/main/java/org/breedinginsight/services/TraitService.java [hunk 3]
```diff
@@ -59,6 +59,8 @@
     private DSLContext dsl;
     private TraitValidatorError traitValidatorError;
 
+    private String FAVORITES_TAG = "favorites";
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/services/TraitService.java:62]: could be final static

File: src/main/java/org/breedinginsight/services/TraitService.java [hunk 4]
```diff
@@ -446,4 +448,21 @@ public Trait setTraitActiveStatus(UUID programId, UUID traitId, Boolean active,
...
+
+        // Parse out tags into unique list
+        Set<String> tags = new HashSet<>();
+        for (Trait trait: traits) {
+            if (trait.getTags() != null) {
+                tags.addAll(trait.getTags());
+            }
+        }
+
+        tags.add(FAVORITES_TAG);
```
Comment (nickpalladino) [inline src/main/java/org/breedinginsight/services/TraitService.java:465]: What is the purpose of this favorites tag?

Comment (ctucker3) [reply src/main/java/org/breedinginsight/services/TraitService.java:465]: There should always be a favorites tag as an option. This sends the favorites tag back in the options of tags even if a trait isn't assigned that tag yet.

**Review by nickpalladino — APPROVED** [review]

**Review by dmeidlin — APPROVED** [review]

**Pull request merged**